File: fleetflow/static/src/js/fleet_flow_dashboard.js
/** @odoo-module **/

import { Component, onWillStart, useState } from "@odoo/owl";
import { registry } from "@web/core/registry";
import { useService } from "@web/core/utils/hooks";
import { rpc } from "@web/core/network/rpc";
import { standardActionServiceProps } from "@web/webclient/actions/action_service";

class FleetFlowDashboard extends Component {
    static template = "fleetflow.Dashboard";
    static props = { ...standardActionServiceProps };

    setup() {
        this.action = useService("action");
        this.state = useState({
            vehicles: {},
            drivers: {},
            trips: {},
            maintenance: {},
            kpis: {},
            isLoading: true,
        });

        onWillStart(async () => {
            await this._fetchDashboardData();
        });
    }

    async _fetchDashboardData() {
        this.state.isLoading = true;
        const data = await rpc("/fleetflow/dashboard/data", {});
        Object.assign(this.state, data, { isLoading: false });
    }

    async refresh() {
        await this._fetchDashboardData();
    }

    // ---- Navigation helpers ----
    openActiveFleet() {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "Active Fleet (On Trip)",
            res_model: "fleet.flow.vehicle",
            view_mode: "kanban,list,form",
            views: [[false, "kanban"], [false, "list"], [false, "form"]],
            domain: [["status", "=", "on_trip"]],
        });
    }

    openMaintenanceAlerts() {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "Vehicles In Shop",
            res_model: "fleet.flow.vehicle",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["status", "=", "in_shop"]],
        });
    }

    openPendingCargo() {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "Pending Cargo",
            res_model: "fleet.flow.trip",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["state", "=", "draft"]],
        });
    }

    openAllVehicles() {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "All Vehicles",
            res_model: "fleet.flow.vehicle",
            view_mode: "kanban,list,form",
            views: [[false, "kanban"], [false, "list"], [false, "form"]],
        });
    }

    openAvailableVehicles() {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "Available Vehicles",
            res_model: "fleet.flow.vehicle",
            view_mode: "kanban,list,form",
            views: [[false, "kanban"], [false, "list"], [false, "form"]],
            domain: [["status", "=", "available"]],
        });
    }

    openActiveTrips() {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "Active Trips",
            res_model: "fleet.flow.trip",
            view_mode: "kanban,list,form",
            views: [[false, "kanban"], [false, "list"], [false, "form"]],
            domain: [["state", "in", ["dispatched", "on_trip"]]],
        });
    }

    openCompletedTrips() {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "Completed Trips",
            res_model: "fleet.flow.trip",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["state", "=", "completed"]],
        });
    }

    openAllDrivers() {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "Drivers",
            res_model: "fleet.flow.driver",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
        });
    }

    openExpiredLicenses() {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "Expired Licenses",
            res_model: "fleet.flow.driver",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["is_license_expired", "=", true]],
        });
    }

    openExpiringSoon() {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "License Expiring Soon",
            res_model: "fleet.flow.driver",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["is_license_expired", "=", false], ["license_days_remaining", "<=", 30]],
        });
    }

    openPendingMaintenance() {
        this.action.doAction({
            type: "ir.actions.act_window",
            name: "Pending Maintenance",
            res_model: "fleet.flow.maintenance",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["state", "in", ["planned", "in_progress"]]],
        });
    }
}

registry.category("actions").add("fleetflow_dashboard", FleetFlowDashboard);

File: fleetflow/controllers/fleet_flow_dashboard.py
from odoo import http
from odoo.http import request


class FleetFlowDashboardController(http.Controller):

    @http.route('/fleetflow/dashboard/data', type='jsonrpc', auth='user')
    def get_dashboard_data(self):
        Vehicle = request.env['fleet.flow.vehicle']
        Driver = request.env['fleet.flow.driver']
        Trip = request.env['fleet.flow.trip']
        Maintenance = request.env['fleet.flow.maintenance']

        total_vehicles = Vehicle.search_count([])
        on_trip_vehicles = Vehicle.search_count([('status', '=', 'on_trip')])
        available_vehicles = Vehicle.search_count([('status', '=', 'available')])
        in_shop_vehicles = Vehicle.search_count([('status', '=', 'in_shop')])
        retired_vehicles = Vehicle.search_count([('status', '=', 'retired')])

        total_drivers = Driver.search_count([])
        on_trip_drivers = Driver.search_count([('status', '=', 'on_trip')])
        on_duty_drivers = Driver.search_count([('status', '=', 'on_duty')])
        suspended_drivers = Driver.search_count([('status', '=', 'suspended')])
        expired_licenses = Driver.search_count([('is_license_expired', '=', True)])
        expiring_soon = Driver.search_count([
            ('is_license_expired', '=', False),
            ('license_days_remaining', '<=', 30),
        ])

        draft_trips = Trip.search_count([('state', '=', 'draft')])
        active_trips = Trip.search_count([('state', 'in', ['dispatched', 'on_trip'])])
        completed_trips = Trip.search_count([('state', '=', 'completed')])

        pending_maintenance = Maintenance.search_count([
            ('state', 'in', ['planned', 'in_progress']),
        ])

        utilization_rate = (
            round(on_trip_vehicles / total_vehicles * 100, 1)
            if total_vehicles else 0
        )

        return {
            'vehicles': {
                'total': total_vehicles,
                'on_trip': on_trip_vehicles,
                'available': available_vehicles,
                'in_shop': in_shop_vehicles,
                'retired': retired_vehicles,
            },
            'drivers': {
                'total': total_drivers,
                'on_trip': on_trip_drivers,
                'on_duty': on_duty_drivers,
                'suspended': suspended_drivers,
                'expired_licenses': expired_licenses,
                'expiring_soon': expiring_soon,
            },
            'trips': {
                'draft': draft_trips,
                'active': active_trips,
                'completed': completed_trips,
            },
            'maintenance': {
                'pending': pending_maintenance,
            },
            'kpis': {
                'active_fleet': on_trip_vehicles,
                'utilization_rate': utilization_rate,
                'maintenance_alerts': in_shop_vehicles,
                'pending_cargo': draft_trips,
            },
        }
